### PR TITLE
Add standalone comment registry

### PR DESCRIPTION
**Pull request by nventuro**
Supercedes https://github.com/balancer-labs/balancer-v2-monorepo/pull/1080, based on the ideas from [this comment](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1080#issuecomment-1021332571) by @TomAFrench.

### CODE REVIEW

**Review by EndymionJkb — COMMENTED** [review]
Added test for getVault(), for completeness.

Another side effect of storing the Vault is we would have separate comment repositories for each deployed Vault (e.g., if we had a "Pro" Vault alongside the current one, with different pools, and probably very different kinds of comments)

**Review by TomAFrench — COMMENTED** [review]
Looks good to me. Couple of thoughts on this:

a)
One thing I'd note is that filtering for events manually tends to be hit and miss as it's slow and RPC providers tend to limit how far back you can look so it's worth having a service which will pull out these comments to store them.

b)
We might want to have an authenticated function for Balancer DAO to make comments from `0xba1...ba1`. This'll allow us to have "official" comments without having reliance on a particular address, and having to migrate if this is compromised. This could also be a separate contract however (except it won't be from `0xba1...ba1`)

c)
I think there's unstated assumption here that we're going to be passing simple strings to this registry however more likely is we're going to want to pass some structured data, e.g. A product built for DAO treasury pools will want to describe which DAO a pool's for, their website, discord, etc. We then need a clean way to encode this.

I've worked on a similar system in the past which emitted the "comment" as a plain bytes array and decode the comment depending on a version number which the user prepended to the comment.

File: pkg/standalone-utils/contracts/PoolCommentRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    function addPoolIdComment(bytes32 poolId, string memory comment) external {
+        _addPoolComment(poolId, comment);
+    }
+
+    function addPoolComment(address pool, string memory comment) external {
+        _addPoolComment(IBasePool(pool).getPoolId(), comment);
+    }
+
+    function _addPoolComment(bytes32 poolId, string memory comment) private {
```
[the remark this answers is not in the document]
Comment (EndymionJkb) [reply pkg/standalone-utils/contracts/PoolCommentRegistry.sol:49]: I'm a little worried there might be regulatory issues around "official" comments, though maybe being from the DAO (vs BLabs) mitigates that somewhat?

I still think that's too much, though. Any official comments could just be on the UI (e.g., through the asset repo mechanism we use for metadata like that).

We could have a service (as described in "a"), that filters/collates them, and the "official" registry is whatever that thing is reading from. So if we ever had to migrate, we could just switch the pointer there. The UI could read from that service, if we wanted to display a "trollbox" for pools.

**Comment by TomAFrench** [comment]
>I'm a little worried there might be regulatory issues around "official" comments, though maybe being from the DAO (vs BLabs) mitigates that somewhat?
>
>I still think that's too much, though. Any official comments could just be on the UI (e.g., through the asset repo mechanism we use for metadata like that).

Agreed that it might be bloating the scope of this contract a bit. If we needed this (and couldn't maintain these comments offchain) then we can just do the separate contract method so no need to add it now.

**Comment by nventuro** [comment]
a) (RPC limitations) Should we provide a way to write to storage as well? It'd be way more expensive, but retrievable in a single RPC call (assuming sufficient gas). We could even do sender and pool filtering on the view function. The only thing I'm not sure about is whether it'd be cheaper to keep those as storage or as contract code - I imagine past a certain length code would be cheaper.

b) (special sender) This seems a bit overkill IMO, we could simply use a multisig or a forwarding contract and it'd work fine. I don't think we need the special hardcoded address. If we ever had to migrate addresses it wouldn't be the end of the world to search for two multiple senders.

c) (format) What about JSON? That seems much more natural.

**Comment by nventuro** [comment]
According to [this](https://github.com/0xsequence/sstore2#sstore-cost-data-writes), writing data as contract code will be cheaper if the data is 32 bytes or longer. There'd be some additional overhead to keep track of all of the data, but 32 seems low enough that that's what I'd do if we were to do this.

**Comment by TomAFrench** [comment]
a) I don't think we should store this onchain. It's trivial to have a subgraph which just hoovers all these up and stores them along with the tx hashes in case someone want to verify a comment onchain. Putting these into storage is overkill.

b) agreed

c) Yeah the first version of data I went with on that project was JSON converted to a string encoded into a byte array. With versioning the data you get a few extra benefits however:
- You can define specs for multiple valid structures of JSON (defined by the version number) and reject anything that doesn't match it rather than allowing people to just throw any JSON blob at you and having to figure out what's in it.
- JSON is reasonably verbose and there may be situations where you want a more concise format (e.g. on L2).

Essentially we'll be able to accept more varied forms of data and it'll be easier to decode.

**Comment by nventuro** [comment]
`{ "version": 2 }` :stuck_out_tongue: 

Valid point however, a `bytes` blob will be more versatile.

**Comment by nventuro** [comment]
Putting this on hold as we think some more about use cases and requirements.

**Pull request closed without merging**